### PR TITLE
add tab-complete to repl

### PR DESCRIPTION
**Pull request by deoxxa**

### CODE REVIEW

**Comment by stevenh** [comment]
Can we get a proper PR description for this @deoxxa it seems you have two commits and you only mention the title from the first one.

**Comment by deoxxa** [comment]
What would you suggest instead? The first commit is only there to support the second one.

**Comment by stevenh** [comment]
Something like:
```Add tab-complete to repl

Add support for using tab complete when using repl.

This also adds the KeysByParent to otto objects which ....```

Subject is typically a brief summary of the change and description provides the detail.

Given the recursive nature of KeysByParent could this cause significant slowdown when not desired?

If would good to add a command line option to disable it.

**Comment by deoxxa** [comment]
Right - can do! Please hold :)

File: otto.go [hunk 1]
```diff
@@ -655,6 +655,25 @@ func (self Object) Keys() []string {
 	return keys
 }
 
+// Get the keys (and those of the parents) for the object, in order of
+// "closest" to "furthest"
+func (self Object) KeysByParent() [][]string {
```
Comment (stevenh) [inline otto.go]: Would be nice if we started to pass linting with doc comments e.g. this should be of the form:
```go
// KeysByParent returns the keys for the object, and those if its parents, in order of "closest" to "furthest"
```

Comment (deoxxa) [inline otto.go]: Can't think of a reason why not!

Comment (deoxxa) [inline otto.go]: Done and done.

**Comment by deoxxa** [comment]
Autocomplete is now opt-in. To avoid adding another set of `OneAndTwoAndThree` functions in the `repl` package, I added an `Options` struct type. The old functions still exist, and just map through to the `RunWithOptions` function, setting the equivalent fields on that argument.

**Pull request merged**